File: packages/api_client/lib/src/api_exception.dart
/// Hata modeli — `docs/03-api-sozlesmesi.md` §1.2.
///
/// Sunucu **tüm** hataları tek biçimde döner. Bu dosya o biçimi Dart tarafında
/// tek bir istisna tipine indirger; uygulamalar HTTP durum koduna değil
/// [ApiErrorCode]'a bakar.
library;

/// Sözleşmedeki hata kodları.
enum ApiErrorCode {
  unauthenticated('UNAUTHENTICATED'),
  forbidden('FORBIDDEN'),
  notFound('NOT_FOUND'),
  validationFailed('VALIDATION_FAILED'),
  invalidTransition('INVALID_TRANSITION'),
  locationClosed('LOCATION_CLOSED'),
  itemUnavailable('ITEM_UNAVAILABLE'),
  deviceRevoked('DEVICE_REVOKED'),
  rateLimited('RATE_LIMITED'),
  serverError('SERVER_ERROR'),

  /// Sunucu sözleşmede olmayan bir kod gönderdi ya da yanıt hiç ayrıştırılamadı
  /// (ağ kesintisi, HTML hata sayfası, zaman aşımı).
  unknown('UNKNOWN');

  const ApiErrorCode(this.wireName);

  final String wireName;

  static ApiErrorCode parse(String? value) {
    if (value == null) return unknown;
    for (final code in ApiErrorCode.values) {
      if (code.wireName == value) return code;
    }
    return unknown;
  }
}

/// API'den dönen her hata bu tipe çevrilir.
class ApiException implements Exception {
  const ApiException({
    required this.code,
    required this.message,
    this.statusCode,
    this.details,
  });

  /// Ağ katmanı hiç yanıt üretemedi (bağlantı yok, zaman aşımı).
  const ApiException.network([String? message])
    : code = ApiErrorCode.unknown,
      message = message ?? 'Bağlantı kurulamadı. İnternetinizi kontrol edin.',
      statusCode = null,
      details = null;

  final ApiErrorCode code;

  /// Kullanıcıya doğrudan gösterilebilir Türkçe metin. Sunucu üretir.
  final String message;

  final int? statusCode;

  /// `VALIDATION_FAILED` için alan bazlı ayrıntı.
  final Map<String, dynamic>? details;

  /// Sözleşmedeki `{ "error": { ... } }` gövdesinden üretir.
  ///
  /// Gövde beklenen biçimde değilse [ApiErrorCode.unknown] ile döner — bir
  /// hata yanıtını ayrıştıramamak, ayrı bir çökme sebebi olmamalı.
  factory ApiException.fromResponse(Object? body, int? statusCode) {
    if (body is! Map) {
      return ApiException(
        code: ApiErrorCode.unknown,
        message: 'Beklenmeyen bir hata oluştu.',
        statusCode: statusCode,
      );
    }

    final error = body['error'];
    if (error is! Map) {
      return ApiException(
        code: ApiErrorCode.unknown,
        message: 'Beklenmeyen bir hata oluştu.',
        statusCode: statusCode,
      );
    }

    final rawDetails = error['details'];
    return ApiException(
      code: ApiErrorCode.parse(error['code'] as String?),
      message: (error['message'] as String?) ?? 'Beklenmeyen bir hata oluştu.',
      statusCode: statusCode,
      details: rawDetails is Map
          ? Map<String, dynamic>.from(rawDetails)
          : null,
    );
  }

  /// Cihaz token'ı iptal edilmiş — KDS eşleme ekranına dönmeli
  /// (`docs/05-mutfakapp.md` §7).
  bool get isDeviceRevoked => code == ApiErrorCode.deviceRevoked;

  /// Oturum düşmüş — istemci giriş ekranına dönmeli.
  bool get isUnauthenticated => code == ApiErrorCode.unauthenticated;

  /// Yeniden denemek anlamlı mı?
  ///
  /// Doğrulama ve yetki hataları tekrar denemekle düzelmez; ağ hatası ve
  /// `429`/`5xx` düzelir. KDS'in üstel geri çekilmesi bunu kullanır.
  bool get isRetryable =>
      code == ApiErrorCode.unknown ||
      code == ApiErrorCode.rateLimited ||
      code == ApiErrorCode.serverError;

  @override
  String toString() => 'ApiException(${code.wireName}: $message)';
}

File: platform/extensions/veykemtu/bridgeapi/src/Exceptions/ApiException.php
<?php

declare(strict_types=1);

namespace Veykemtu\BridgeApi\Exceptions;

use Exception;
use Illuminate\Http\JsonResponse;

/**
 * Sözleşmedeki tek hata biçimi — `docs/03-api-sozlesmesi.md` §1.2.
 *
 * Tüm API hataları bu tipe indirgenir; istemciler HTTP durum koduna değil
 * `error.code` alanına bakar. `message` doğrudan kullanıcıya gösterilebilir
 * Türkçe metindir.
 */
class ApiException extends Exception
{
    /** @param array<string, mixed>|null $details */
    public function __construct(
        public readonly string $errorCode,
        string $message,
        public readonly int $status,
        public readonly ?array $details = null,
    ) {
        parent::__construct($message);
    }

    public function render(): JsonResponse
    {
        return new JsonResponse([
            'error' => array_filter([
                'code' => $this->errorCode,
                'message' => $this->getMessage(),
                'details' => $this->details,
            ], static fn($value): bool => $value !== null),
        ], $this->status, [], JSON_UNESCAPED_UNICODE);
    }

    public static function unauthenticated(
        string $message = 'Oturum bulunamadı, tekrar giriş yapın.',
    ): self {
        return new self('UNAUTHENTICATED', $message, 401);
    }

    public static function forbidden(
        string $message = 'Bu uca erişim yetkiniz yok.',
    ): self {
        return new self('FORBIDDEN', $message, 403);
    }

    public static function deviceRevoked(): self
    {
        return new self(
            'DEVICE_REVOKED',
            'Bu cihazın yetkisi iptal edilmiş. Yönetici ile görüşün.',
            403,
        );
    }

    public static function notFound(string $message = 'Kayıt bulunamadı.'): self
    {
        return new self('NOT_FOUND', $message, 404);
    }

    /** @param array<string, mixed>|null $details */
    public static function validationFailed(string $message, ?array $details = null): self
    {
        return new self('VALIDATION_FAILED', $message, 422, $details);
    }

    public static function invalidTransition(string $from, string $to, string $reason): self
    {
        return new self('INVALID_TRANSITION', $reason, 422, [
            'from' => $from,
            'to' => $to,
        ]);
    }

    public static function locationClosed(
        string $message = 'Şu anda sipariş alınmıyor.',
    ): self {
        return new self('LOCATION_CLOSED', $message, 422);
    }

    public static function itemUnavailable(string $message, int $menuId): self
    {
        return new self('ITEM_UNAVAILABLE', $message, 422, ['menu_id' => $menuId]);
    }

    public static function serverError(
        string $message = 'Beklenmeyen bir hata oluştu.',
    ): self {
        return new self('SERVER_ERROR', $message, 500);
    }
}
